### PR TITLE
bulk add opportunities

### LINKED ISSUE #338: CSV Upload Opportunities
First-time users need a way to bulk-add opportunities, on the way to users having real data from which to generate a map (#227; note that #201 is in between this and that). We're thinking CSV for now, but it has to be snappy!

### PR DESCRIPTION
**Pull request by chadwhitacre**
Closes #338.
#### To-do
- [x] add a little inline documentation of the format
- [x] add a form field to select the organizer based on available organizers for logged-in user
- [x] implement idempotency on `name`
- [x] provide detailed tabular error reporting

### CODE REVIEW

**Comment by dmtroyer** [comment]
http://guides.rubyonrails.org/form_helpers.html

**Comment by chadwhitacre** [comment]
!m @dmtroyer

**Comment by chadwhitacre** [comment]
Over to you, @dmtroyer! :-)

**Comment by dmtroyer** [comment]
image

**Comment by chadwhitacre** [comment]
!m @dmtroyer

**Comment by dmtroyer** [comment]
Currently, the csv parsing logic expects a header in the csv which defines the names of the fields being uploaded in the order listed in the header.

```
name,description
Marbles,"David's got marbles"
"Skeet shooting","Shoot some skeet"
```

**Comment by chadwhitacre** [comment]
Getting back into this, hitting `undefined method `id' for nil:NilClass` in `opportunities_controller.rb:39`:

``` ruby
      Opportunity.transaction do
        CSV.parse(params[:csv].read, headers: true) do |row|
          row['resource_sub_type_id'] = ResourceSubType.find_by_name(row.delete('resource_sub_type')).id
          @opportunities.push(Opportunity.create!(row.to_h))
        end
      end
```

**Comment by chadwhitacre** [comment]
That's with the test data from above at https://github.com/saxifrage/cityasacampus/pull/341#issuecomment-154545562. Looks like 266ce9127d76a842a2db07878cbc5e829c3569a2 added an expectation for the CSV structure ...

**Comment by chadwhitacre** [comment]
Blam!

``` csv
name,description,resource_sub_type
Marbles,"David's got marbles",Event
"Skeet shooting","Shoot some skeet",Event
```

screen shot 2015-11-09 at 9 53 57 am

**Comment by chadwhitacre** [comment]
(For my notes: I found the database config in `config/database.yml`, and then found the available resource subtypes like so:

``` sql
learn_dev=# select * from resource_sub_types;
┌────┬───────┬──────────────────┐
│ id │ name  │ resource_type_id │
├────┼───────┼──────────────────┤
│  1 │ Event │                1 │
└────┴───────┴──────────────────┘
(1 row)

learn_dev=#
```

)

**Comment by chadwhitacre** [comment]
screen shot 2015-11-09 at 10 01 22 am

**Comment by chadwhitacre** [comment]
I've started a to-do list in the PR description.

**Comment by dmtroyer** [comment]
There are some fields that are required, `resource_sub_type` was the only one that was actually being validated so that's where I started, and I chose to use the name in the csv. A logged on user can have more than one organizer, so they'll probably have to be explicit or select the organizer as part of the upload form or something.

**Comment by chadwhitacre** [comment]
I'm trying to understand the Opportunity model and the opportunities table. They don't appear to sync up. The latter has way more fields than the former.

``` ruby
class Opportunity < ActiveRecord::Base
  has_many :opportunity_instances
  belongs_to :organizer
  has_one :after_this, class_name: "Opportunity", foreign_key: "after_this_id"
  has_one :before_this, class_name: "Opportunity", foreign_key: "before_this_id"
  belongs_to :resource_sub_type
  has_one :resource_type, through: :resource_sub_type

  validates :resource_sub_type, presence: true

  def active_model_serializer
    OpportunitySerializer
  end
end
```

``` sql
Null display is "¤".
Line style is unicode.
Border style is 2.
                                            Table "public.opportunities"
┌───────────────────────┬─────────────────────────────┬────────────────────────────────────────────────────────────┐
│        Column         │            Type             │                         Modifiers                          │
├───────────────────────┼─────────────────────────────┼────────────────────────────────────────────────────────────┤
│ id                    │ integer                     │ not null default nextval('opportunities_id_seq'::regclass) │
│ name                  │ character varying           │                                                            │
│ address               │ text                        │                                                            │
│ description           │ text                        │                                                            │
│ registration_url      │ character varying           │                                                            │
│ location_name         │ character varying           │                                                            │
│ registration_deadline │ timestamp without time zone │                                                            │
│ program_type          │ character varying           │                                                            │
│ logo_url              │ character varying           │                                                            │
│ starts_at             │ timestamp without time zone │                                                            │
│ ends_at               │ timestamp without time zone │                                                            │
│ online_address        │ character varying           │                                                            │
│ zipcode               │ character varying           │                                                            │
│ city                  │ character varying           │                                                            │
│ state                 │ character varying           │                                                            │
│ is_online             │ boolean                     │                                                            │
│ hide_reason           │ character varying           │                                                            │
│ hide                  │ boolean                     │                                                            │
│ contact_name          │ character varying           │                                                            │
│ contact_email         │ character varying           │                                                            │
│ contact_phone         │ character varying           │                                                            │
│ price_level           │ integer                     │                                                            │
│ min_age               │ integer                     │                                                            │
│ max_age               │ integer                     │                                                            │
│ extra_data            │ text                        │                                                            │
│ created_at            │ timestamp without time zone │ not null                                                   │
│ updated_at            │ timestamp without time zone │ not null                                                   │
│ organizer_id          │ integer                     │                                                            │
│ topic_id              │ integer                     │                                                            │
│ after_this_id         │ integer                     │                                                            │
│ before_this_id        │ integer                     │                                                            │
│ badge_class_id        │ character varying           │                                                            │
│ resource_sub_type_id  │ integer                     │                                                            │
└───────────────────────┴─────────────────────────────┴────────────────────────────────────────────────────────────┘
Indexes:
    "opportunities_pkey" PRIMARY KEY, btree (id)
    "index_opportunities_on_after_this_id" btree (after_this_id)
    "index_opportunities_on_before_this_id" btree (before_this_id)
    "index_opportunities_on_organizer_id" btree (organizer_id)
    "index_opportunities_on_resource_sub_type_id" btree (resource_sub_type_id)
    "index_opportunities_on_topic_id" btree (topic_id)
Foreign-key constraints:
    "fk_rails_6d4b8f5561" FOREIGN KEY (topic_id) REFERENCES topics(id)
```

**Comment by chadwhitacre** [comment]
> A logged on user can have more than one organizer, so they'll probably have to be explicit or select the organizer as part of the upload form or something.

Okay, noted in to-do.

**Comment by chadwhitacre** [comment]
@dmtroyer Do we have a test suite? How do I run it?

**Comment by chadwhitacre** [comment]
```
$ bundle exec rake test
[...]
35 runs, 0 assertions, 0 failures, 35 errors, 0 skips
```

**Comment by dmtroyer** [comment]
> I'm trying to understand the Opportunity model and the opportunities table. They don't appear to sync up. The latter has way more fields than the former.

Rails is cheeky like that, it dynamically creates default getter and setters for attributes defined in the schema. So, if you're cool with the defaults they don't have to be explicitly defined in the model.

**Comment by chadwhitacre** [comment]
Okay, thanks for the pointer.

```
$ head -n12 db/schema.rb
# encoding: UTF-8
# This file is auto-generated from the current state of the database. Instead
# of editing this file, please use the migrations feature of Active Record to
# incrementally modify your database, and then regenerate this schema definition.
#
# Note that this schema.rb definition is the authoritative source for your
# database schema. If you need to create the application database on another
# system, you should be using db:schema:load, not running all the migrations
# from scratch. The latter is a flawed and unsustainable approach (the more migrations
# you'll amass, the slower it'll run and the greater likelihood for issues).
#
# It's strongly recommended that you check this file into your version control system.
```

**Comment by chadwhitacre** [comment]
> There are some fields that are required

How is this information encoded? I only see three out of 33 columns in the schema that are `not null`, and I don't see anything suggestive of requirement in `opportunity.rb`.

**Comment by dmtroyer** [comment]
> How is this information encoded? I only see three out of 33 columns in the schema that are not null, and I don't see anything suggestive of requirement in opportunity.rb.

A couple of different ways:
1. At the database and schema level through [migrations](http://edgeguides.rubyonrails.org/active_record_migrations.html)
2. At the model level through [validations](http://guides.rubyonrails.org/active_record_validations.html) https://github.com/saxifrage/cityasacampus/blob/master/app/models/opportunity.rb#L9

**Comment by chadwhitacre** [comment]
@dmtroyer I only see one validated (required) field in `opportunity.rb`. In `schema.rb` there are only two `null: false` fields (there is of course an `id` field in the actual database schema that is also `not null`): `created_at` and `updated_at`. Are those the only three (or four) required fields, then?

The backtrace at https://github.com/saxifrage/cityasacampus/pull/341#issuecomment-155087940 seems to indicate that `organizer_id` ought to be required as well, and that therefore in general we are not yet requiring all of the fields that we should be.

**Comment by dmtroyer** [comment]
@whit537 All correct. Field requirement and validation has not been completely flushed out.

**Comment by chadwhitacre** [comment]
Okay!

**Comment by chadwhitacre** [comment]
I'm gonna pull @timothyfcook over to make some decisions about required fields for the CSV upload.

**Comment by dmtroyer** [comment]
!m @whit537 @timothyfcook

**Comment by chadwhitacre** [comment]
The word from @timothyfcook:

```
name
description
registration_url
program_type
logo_url
is_online
contact_name
contact_email
contact_phone
price_level
min_age
max_age
organizer_id
topic_id
resource_sub_type_id
```

If `is_online`, then `online_address` should be required. Otherwise, `address`, `city`, `state`, and `zipcode` should be required.

**Comment by chadwhitacre** [comment]
screen shot 2015-11-09 at 12 24 40 pm

**Comment by chadwhitacre** [comment]
Aren't there supposed to be gems for everything? Isn't that why we use Rails? Let's go look! :-)

https://www.ruby-toolbox.com/categories/CSV_Parsers

**Comment by dmtroyer** [comment]
@whit537 there is the [Ruby CSV Standard Library](http://ruby-doc.org/stdlib-2.0.0/libdoc/csv/rdoc/CSV.html) that I started using, although [FasterCSV](https://www.ruby-toolbox.com/projects/fastercsv) claims to be, well, faster. What functionality are you looking for?

**Comment by chadwhitacre** [comment]
@dmtroyer I was looking for something higher-order that provided the validation of all rows that we're looking for. FasterCSV looks pretty mature, but I'm not sure it adds value for us.

**Comment by chadwhitacre** [comment]
Looks like FasterCSV _is_ the stdlib's CSV as of Ruby 1.9:

http://graysoftinc.com/my-projects/getting-fastercsv-ready-for-ruby-19

**Comment by chadwhitacre** [comment]
Okay, not finding anything interesting. Gonna proceed to write our own stuff here.

**Comment by chadwhitacre** [comment]
What if we required `online_address` in all cases, and used that as the `ref` for idempotency?

**Comment by chadwhitacre** [comment]
We could special-case LearnPgh.org `online_address`es for subsequent uploads. That would allow folks to modify their records that don't otherwise have a web page somewhere and us to still use `online_address` for idempotency.

**Comment by dmtroyer** [comment]
Modifying/updating records seems to me like it could be a second phase of this uploader and a place to save time. I would think users could use the rest of the admin dashboard to edit single opportunities.

As far as idempotency goes, what's the expected behavior when uploading a dup? Swallow it? I'm not following the `online_address` logic.

**Comment by chadwhitacre** [comment]
Here's some bare-bones error handling:

screen shot 2015-11-09 at 2 57 14 pm

**Comment by chadwhitacre** [comment]
I tried using some lambdas but apparently I don't know what I'm doing. :-/

``` ruby
          deref = -> (x, y) { x.find_by_name(row.delete[y]) }
          xid = -> (x) { unless x.nil? then x.id else nil end }

          row['organizer_id'] = xid(deref(Organizer, 'organizer'))
          row['topic_id'] = xid(deref(Topic, 'topic'))
          row['resource_sub_type_id'] = xid(deref(ResourceSubType, 'resource_sub_type'))
```

screen shot 2015-11-09 at 3 36 03 pm

**Comment by chadwhitacre** [comment]
/me gives up on that approach ...

**Comment by chadwhitacre** [comment]
Here's some finer-grained error handling:

screen shot 2015-11-09 at 3 52 49 pm

**Comment by chadwhitacre** [comment]
``` csv
name,description,registration_url,program_type,logo_url,is_online,contact_name,contact_email,contact_phone,price_level,min_age,max_age,organizer,topic,resource_sub_type
Marbles,"David's got marbles",http://example.com/,Program,http://example.com/logo,true,Alice,alice@example.com,5558675309,10,0,,Assemble,Making,Event
"Skeet shooting","Shoot some skeet",,,,,,,,,,,,,
```

**Comment by chadwhitacre** [comment]
screen shot 2015-11-09 at 4 00 46 pm

**Comment by chadwhitacre** [comment]
screen shot 2015-11-09 at 4 33 28 pm

**Comment by chadwhitacre** [comment]
screen shot 2015-11-09 at 4 37 31 pm

File: app/controllers/opportunities_controller.rb [hunk 1]
```diff
@@ -29,6 +29,44 @@ def new
     @opportunity = Opportunity.new
   end
 
+  @@permitted_params = [:name, :address, :description, :registration_url, :location_name,
```
Comment (chadwhitacre) [inline app/controllers/opportunities_controller.rb:32]: Is this the right way to do this, @dmtroyer? I'm trying to expose the `permitted_params` to the view so we can loop over them programmatically.

**Comment by chadwhitacre** [comment]
Good data works!

screen shot 2015-11-09 at 4 44 27 pm

``` csv
name,description,registration_deadline,starts_at,ends_at,registration_url,program_type,logo_url,is_online,contact_name,contact_email,contact_phone,price_level,min_age,max_age,organizer,topic,resource_sub_type
Marbles,"David's got marbles",2020-10-10,2020-10-10,2020-10-10,http://example.com/,Program,http://example.com/logo,true,Alice,alice@example.com,5558675309,10,0,1000,Assemble,Making,Event
"Skeet shooting","Shoot some skeet",,,,,,,,,,,,,,,,
```

**Comment by chadwhitacre** [comment]
Okay! Basic framework is roughed in. Now what?

**Comment by chadwhitacre** [comment]
Why are none of these showing up on the main Opportunities listing on the dashboard? What is [`policy_scope`](https://github.com/saxifrage/cityasacampus/blob/77f8f9138aaa844681aa1292c931b05822cc8a67/app/controllers/opportunities_controller.rb#L11)?

screen shot 2015-11-09 at 5 21 17 pm

**Comment by chadwhitacre** [comment]
https://github.com/elabs/pundit#scopes

**Comment by chadwhitacre** [comment]
Looks like we [constrain that listing by organizer](https://github.com/saxifrage/cityasacampus/blob/77f8f9138aaa844681aa1292c931b05822cc8a67/app/policies/opportunity_policy.rb).

**Comment by chadwhitacre** [comment]
So let's get organizer into the bulk adder! :-)

**Comment by chadwhitacre** [comment]
Blam!

screen shot 2015-11-09 at 6 27 40 pm

**Comment by chadwhitacre** [comment]
> add a `ref` to opportunities to support idempotency

@timothyfcook and I decided irl conversation to use `name` for idempotency: it is required and should be unique per `organizer`.

**Comment by chadwhitacre** [comment]
Didn't quite land duplicate prevention. :-/

File: app/models/opportunity.rb [hunk 2]
```diff
@@ -8,6 +8,24 @@ class Opportunity < ActiveRecord::Base
...
+  validates :contact_phone,     presence: true, allow_blank: false
+
+  validates :price_level,       presence: true, :numericality => { :greater_than_or_equal_to => 0 }
+  validates :min_age,           presence: true, :numericality => { :greater_than_or_equal_to => 0 }
+  validates :max_age,           presence: true, :numericality => { :greater_than_or_equal_to => 0 }
+
+  validates :registration_deadline, presence: true
+  validates :starts_at,         presence: true
+  validates :ends_at,           presence: true
+
```
Comment (MatthewVita) [inline app/models/opportunity.rb:29]: dat validation :+1:

**Comment by dmtroyer** [comment]
@whit537 thought I'd help out a little with the uniqueness constraint on `name` per `organizer`. [Active Record Uniqueness Validations](http://guides.rubyonrails.org/active_record_validations.html#uniqueness) save the day here.

**Comment by chadwhitacre** [comment]
Blam. Thanks, @dmtroyer. I was just about to ask you for help! :-)

screen shot 2015-11-09 at 7 56 26 pm

**Comment by chadwhitacre** [comment]
And with that, I'm out for the rest of the week. Heading to New York for https://github.com/gratipay/inside.gratipay.com/issues/384 this weekend. See you next week! :-)

**Comment by dmtroyer** [comment]
@whit537 have fun! :bus: :city_sunset: :dollar:

**Comment by chadwhitacre** [comment]
Rebased on master. Former head was c33b5671b590c485c4aec6c4422ff329f576e620.

**Comment by chadwhitacre** [comment]
screen shot 2015-11-17 at 3 45 02 pm

**Comment by chadwhitacre** [comment]
Working on inline docs.

screen shot 2015-11-23 at 9 20 40 am

**Comment by chadwhitacre** [comment]
I added a dynamic listing of the possible values for `topic` and `resource_sub_type`.

screen shot 2015-11-23 at 9 35 26 am

**Comment by chadwhitacre** [comment]
Okay! I think this is ready to merge.

Waddya think, @dmtroyer @timothyfcook? :-)

**Comment by timothyfcook** [comment]
I'll let @dmtroyer sign off on this one

**Comment by chadwhitacre** [comment]
Waddya say, @dmtroyer? :-)

**Comment by dmtroyer** [comment]
Will do sometime today!

**Comment by dmtroyer** [comment]
oops, unsure what the previous head was. sorry, seems to be OK. I'll probably squash it before I merge it, anyhow...

**Comment by chadwhitacre** [comment]
!m @dmtroyer

**Pull request merged**